File: src/bmsx/machine/devices/audio/pcm_decoder.ts
import {
	APU_FAULT_NONE,
	APU_FAULT_OUTPUT_DATA_RANGE,
	type ApuAudioSource,
} from './contracts';

export type ApuPcmValidationResult = {
	faultCode: number;
	faultDetail: number;
};

const APU_PCM_VALIDATION_OK: ApuPcmValidationResult = { faultCode: APU_FAULT_NONE, faultDetail: 0 };

export function validateApuPcmSourceData(source: ApuAudioSource): ApuPcmValidationResult {
	const bytesPerSample = source.bitsPerSample === 16 ? 2 : 1;
	const requiredDataBytes = source.frameCount * source.channels * bytesPerSample;
	if (requiredDataBytes > source.dataBytes) {
		return { faultCode: APU_FAULT_OUTPUT_DATA_RANGE, faultDetail: source.dataBytes };
	}
	return APU_PCM_VALIDATION_OK;
}

File: src/bmsx_cpp/machine/devices/audio/pcm_decoder.cpp
#include "machine/devices/audio/pcm_decoder.h"

namespace bmsx {

namespace {
constexpr ApuPcmValidationResult APU_PCM_VALIDATION_OK{};
} // namespace

ApuPcmValidationResult validateApuPcmSourceData(const ApuAudioSource& source) {
	const u64 bytesPerSample = source.bitsPerSample == 16u ? 2u : 1u;
	const u64 requiredDataBytes = static_cast<u64>(source.frameCount) * static_cast<u64>(source.channels) * bytesPerSample;
	if (requiredDataBytes > static_cast<u64>(source.dataBytes)) {
		return {APU_FAULT_OUTPUT_DATA_RANGE, source.dataBytes};
	}
	return APU_PCM_VALIDATION_OK;
}

} // namespace bmsx
